### PR TITLE
ffmpeg6 split

### PR DESCRIPTION
**Pull request by zlice**
[ci skip]

#### Testing the changes

- I tested the changes in this PR: **YES**

main testing has been vaapi in Firefox Nightly and enc/dec in Shotcut on Intel A770

#### New package

- This new package conforms to the [package requirements](https://github.com/void-linux/void-packages/blob/master/CONTRIBUTING.md#package-requirements): **YES**

#### Local build testing

- I built this PR locally for my native architecture, **x86_64**

#### Discussion

Continuation of https://github.com/void-linux/void-packages/pull/36315

`libavresample` is deprecated, but marked as a dependency when trying to build other packages and install `ffmpeg`(6). I think the right way to handle this is make a meta package and the add to removals?

##### Commit naming / misc

~~Commits are 'dirty' right now, running with [gstreamer 1.22.1 updates](https://github.com/void-linux/void-packages/pull/42861) and [Intel OpenCL](https://github.com/void-linux/void-packages/pull/43084). This will also make [intel packages pr](https://github.com/void-linux/void-packages/pull/41132) obsolete.~~ (see Requirements below)

Maintainer for previous ffmpeg(4) was orphaned, I can put my name in there if that matters, though I don't exactly see myself keeping up on ffmpeg.

~~This updates QT5 which looks like it is usually hosted by Johnny. Not sure if there's any reason to stay there or these sources are fine.~~ Using qt5.15.11 now, revbumps and some patches.

##### STILL FFMPEG4

These are a mix of emulators, games and/or older packages.

- kodi (doesn't build, on version 19 - arch pkgbuild updated jul 4th 2024, looks like version 21 and ffmpeg6)
- tvheadend (download page recommends using 4.3 which compiles fine, 4.4 eta is Q1 2024)
- idjc (0.9.8 compiles, remove python3.10 patch, needs makedepends libebur128-devel and libshout-idjc 2.4.6 with '-r1' for distfile - https://github.com/void-linux/void-packages/pull/48944)
- flowblade/mlt - flowblade 2.12.0.2 with mlt7-python3 can 'render' videos with cpu and gpu - ~~https://github.com/void-linux/void-packages/pull/47195~~ https://github.com/void-linux/void-packages/pull/50385)
- mediastreamer (linphone links are dead, changed to github belladonne. 5.3.63 does not use ffmepg5+ still)
- `vlc` (vlc4 will support ffmpeg5, there's a patch to build 3.x with ffmpeg5 but it will lack vaapi)
- olive (has been in alpha since 2020, seems they do nightly builds and appimages)
- moc
- ~~attract~~ (2.7 builds)
- ~~audacity (https://github.com/audacity/audacity/issues/2305 and https://sourceforge.net/p/portmedia/wiki/portsmf/ is 'beta')~~ (updated to 3.x, compiles with ffmpeg6)
- ~~deadbeef (1.9.6 may work but requires libdispatch from apple)~~ 1.9.6 works
- ~~ccextractor~~ (master patches)
- ~~vice (3.8 compiles, needs makedepends libcurl-devel)~~ ty yosh
- ~~synfig (mlt)~~ (work with mlt7, bump synfigstudio)
- ~~ssr~~ (patches)
- ~~spek-alternative~~ (replace with spek-x)
- ~~ppsspp~~ ~~(1.17.1 is out, but there's 5+ distfiles that are all git commits)~~ (minor patch)
- ~~libextractor (1.13 compiles, remove exiv2 patch)~~ (as yosh pointed out, 1.13 just removes ffmpeg bits)
- ~~hedgewars~~ (patch for 6 https://github.com/hedgewars/hw/pull/74)
- ~~ffmpegthumbnailer~~ (patches for moviedecoder.xxx files)
- ~~ffmpegthumbs~~ (latest compiles)
- ~~omxplayer~~ (package removed, deprecated in favor of vlc)
- ~~`osg` (OpenSceneGraph plus `xine-lib`) (patches from osg issue build with ffmpeg6, [ffmpeg5.patch.txt](https://github.com/void-linux/void-packages/files/14743246/ffmpeg5.patch.txt))~~ used ffmpeg6 patch
- ~~qtav (fork has ffmpeg 5 patch - https://github.com/cmguo/QtAV)~~ qtav was absorbed by digikam, which is the only user https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=270213

##### HAVE NOT BUILT YET

- ~~omxplayer (rpi only? ffmpeg4)~~ (removed)
- kodi (big, but ffmpeg4)

##### UPDATED (for ffmpeg6)

- ~~blender: update to 3.5.1~~ ~~will be revbump with~~ https://github.com/void-linux/void-packages/pull/43992
- unpaper: update to 7.0.0.
- guvcview: update to 2.0.8.
- ~~qt5: update to 5.15.8~~ updated by another PR to 5.15.11
- ~~qt5-webengine: update to 5.18.15~~ updated by another PR to 5.15.16 (some patching for ffmpeg build)
- ~~shiboken2: revbump for qt5.15.8 (patch)~~ updated by another PR to 5.15.10
- ~~python3-pyside2: update to 5.15.8 (patch)~~ updated by another PR to 5.15.10
- New package: spex-x replaces spek-alternative for ffmpeg-6.0

##### Requirements

- ~~[gstreamer 1.22.2](https://github.com/void-linux/void-packages/pull/42861)~~ - a few packages use ffmpeg-devel and require a revbump

- ~~[Intel Packages](https://github.com/void-linux/void-packages/pull/41132) [Intel Packages v2](https://github.com/void-linux/void-packages/pull/47330) - ffmpeg4 uses different libraries than 6 (libmfx / Intel-MediaSDK vs only oneVPL that Intel is moving to), both are included here~~. Using replacement oneVPL from OBS update.

- ~~[shotcut 23.05.14](https://github.com/void-linux/void-packages/pull/44260)~~

- ~~https://github.com/void-linux/void-packages/pull/44781 will remove from this PR~~ https://github.com/void-linux/void-packages/commit/939a712449718b8fa805e31ae781f80e3f014ef7 nv-codec-headers updated but breaks 6.0.1 and 4.4.4

### CODE REVIEW

**Comment by paper42** [comment]
are the commits for a blender update, gstreamer and new intel packages really necessary for this PR? If not, please drop them. If yes and they could be merged before merging this PR, please split them to a separate PR and mention that it should be merged before this PR in this PR's description.

**Comment by zlice** [comment]
**Intel OpenCL** commits removed

**blender** - uses ffmpeg-devel, added to update list

**gstreamer** - added to 'Requirements' section and only revbumped the appropriate packages

**Intel Packages** - added to 'Requirements'

**Comment by zlice** [comment]
ICU updated, and many other things since my last full build test. building fine for now

**Comment by yoshiyoshyosh** [comment]
we can replace `spek-alternative` with [`spek-x`](https://github.com/MikeWang000000/spek-X ), a more actively maintained fork that supports ffmpeg6. [the original `spek`](https://github.com/alexkay/spek) has also become maintained again, but it claims to only support up to ffmpeg 5.1.

**Comment by zlice** [comment]
Blocked by https://github.com/void-linux/void-packages/pull/46172

Already quite a few updates to sift through and I don't feel comfortable trying until I know I can update mesa.

**Comment by SpidFightFR** [comment]
> Blocked by #46172
> 
> Already quite a few updates to sift through and I don't feel comfortable trying until I know I can update mesa.

Sup, 23.1.8 should build, but it doesn't work because 32-bit subpackage problem. i still don't know how to solve this issue.

**Comment by zlice** [comment]
Got 64/32bit x86 building again. tyvm spid

**Comment by SpidFightFR** [comment]
> Got 64/32bit x86 building again. tyvm spid

it's not ready yet, i believe there is an issue with arm and aarch64 archs.
Mesa still can't be updated atm.

**Comment by SpidFightFR** [comment]
see here: https://github.com/void-linux/void-packages/issues/46674

**Comment by zlice** [comment]
> we can replace `spek-alternative` with [`spek-x`](https://github.com/MikeWang000000/spek-X), a more actively maintained fork that supports ffmpeg6. [the original `spek`](https://github.com/alexkay/spek) has also become maintained again, but it claims to only support up to ffmpeg 5.1.

added to here, separate PR was building off ffmpeg4 and looked like everything was bombing out.

not sure if i did the transitional dummy package to replace spek-alternative with spek-x right

**Comment by mhmdanas** [comment]
I'd like to help move this along, is there anything I can do?

**Comment by mhmdanas** [comment]
Thanks for all the info, when I get a bit more time (hopefully in a couple weeks), I'll sift through the items you mentioned. :)

**Comment by zlice** [comment]
Using QT 5.15.11 that was just updated. Revbump for qt5 and some hacky patch to build qt5-webengine by forcing `-system-ffmpeg` and knocking out the test was complaining. If that should be fixed properly lmk.

**Comment by zlice** [comment]
https://github.com/void-linux/void-packages/pull/46013

**Comment by zlice** [comment]
updated comments for ffmpeg4 packages. flowblade seems to work for me and should allow for removal of mlt and can use mlt7.

vice, libextractor and idjc i wouldn't know how to test. mediastreamer is linphone based so i didn't check beyond seeing a new version (complained about optr cmake files?)

**Comment by zlice** [comment]
I'm about to make all revisions 99 until this is about to merge x-x

**Comment by Luciogi** [comment]
add flowblade PR link: https://github.com/void-linux/void-packages/pull/47195

**Comment by yoshiyoshyosh** [comment]
heya, I have enough time that I'd like to help move this along. has much changed in terms of what needs to be done since your last comment about what needs to be done?

**Comment by zlice** [comment]
i don't think much more can be done besides `vice, libextractor and idjc` being tested real quick. `tvheadend` update should hit at some point. nvidia headers could be merged. `mlt (flowblade)` can happen after. new `gonic` package can use ffmpeg6.

really just need to clean up commits and double check things.

as for testing, even slackware has ffmpeg 6.1, so i'm sure package compatibility fine at least for glibc.

**Comment by yoshiyoshyosh** [comment]
vice and idjc work well (well, idjc spews a bunch of syntax parsing warnings but that's python stuff rather than ffmpeg stuff). libextractor's throwing an error with exiv2 and I'm not sure if that's affecting the ffmpeg part or not

**Comment by zlice** [comment]
> vice and idjc work well

@yoshiyoshyosh would you mind making PRs to update vice/idjc version since you tested them? then i'll bump them in here

**Comment by yoshiyoshyosh** [comment]
> @yoshiyoshyosh would you mind making PRs to update vice/idjc version since you tested them? then i'll bump them in here

@zlice sounds good. also, I figured out something with libextractor: it doesn't need ffmpeg anymore with the latest versions. can also make a pr to

**Comment by zlice** [comment]
must be why it built lol. odd, they removed it and didn't replace it with anything so opus and thumbnails are gone? only thing that uses it is `gnunet`. thanks

**Comment by yoshiyoshyosh** [comment]
> so opus and thumbnails are gone?
yup. I think it's just a symptom of being really old and half-abandoned by now haha

**Comment by yoshiyoshyosh** [comment]
also, it seems like the latest idjc versions can build and run with ffmpeg6 perfectly fine.

**Comment by yoshiyoshyosh** [comment]
idk if you did tests for spek-x when building it but they seem to fail for me without this patch:
```diff
--- a/tests/perf.cc	2024-03-09 17:13:49.086603418 -0500
+++ b/tests/perf.cc	2024-03-09 17:13:59.453603753 -0500
@@ -1,3 +1,4 @@
+#include <cstdint>
 #include <cstdlib>
 #include <iostream>
 #include <fstream>
```

**Comment by zlice** [comment]
> idk if you did tests for spek-x when building it but they seem to fail for me without this patch:

haven't done full builds for musl yet. when your PRs get merged i'll probably start bugging people. coming up on a year now and my updates just needed another kick. probably doing a full reinstall once this merges.

**Comment by yoshiyoshyosh** [comment]
> haven't done full builds for musl yet

that's odd, it was failing while I was testing on glibc O.o, wonder if there's some differing strict compiler flags present somewhere. whatever, so long as it works

**Comment by zlice** [comment]
usually when i have to include stdint it's a musl thing. ya builds fine for me on glibc. don't know of anything fancy i do

**Comment by Luciogi** [comment]
@zlice flowblade old PR got messed up, here is new PR https://github.com/void-linux/void-packages/pull/50385

**Comment by zlice** [comment]
> @zlice flowblade old PR got messed up, here is new PR #50385

Awesome. Dang looks like builds aren't going for it D=

**Comment by zlice** [comment]
Made mlt7 SDL2. Their git is tagged to sdl2 though build searches for either. Suppose 1 would have less revbumps but it seems benign.

File: srcpkgs/spek-x/template
Comment (ahesford) [inline srcpkgs/spek-x/template:15]: This should include a `replaces="spek-alternatives<=0.8.2.3_2"` line to make sure upgrades are properly handled.

**Review by ahesford — CHANGES_REQUESTED** [review]
This is a massive and far-reaching PR. I think we should scale it back:
1. Keep `ffmpeg` at its current version (4.4.4). This lets us drop all revbumps that would need to be rebuilt for a new `ffmpeg4`. That's about 20% of the templates you've touched.
2. Make a new package, `ffmpeg6` and any of the new additions necessary to support it.
3. Delete the `freecad` commit, it doesn't seem to belong.
4. The current `ffmpeg` dependants that can be easily migrated to the new `ffmpeg6` can be included in this PR, but let's defer some of the large rebuilds until later. For example:
    - electron19
    - electron24
    - qt6-pdf (this might include qt6-multimedia)
    - qt5-webengine (this might include qt5 and qt5-speech)
 
I'm not sure if there are other interactions between qt stuff that might depend on the old `ffmpeg` and dependants that might move to the new `ffmpeg6`, but those whole branches can remain on the old `ffmpeg` for now. As we update the electron and qt packages, we can move them to `ffmpeg6` and eventually sunset `ffmpeg`.

**Comment by zlice** [comment]
> This is a massive and far-reaching PR. I think we should scale it back:

Updating this or making another PR with a ffmpeg6 package should be easy enough, but I can't really test ffmpeg(4)+ffmpeg6 without a reinstall. Hope to make a new PC when ryzen 9000 drops. From what I remember talking with paper a while back I recall there was something that made switching that way (ffmpeg+ffmpeg6) a no-go or harder to do. Would you make 4 depend on 6 and use the binaries for ffmpeg/ffprobe/etc from ffmpeg6 ? Would that mess with dependencies or keeping a plain 'ffmpeg' package?

It looks closer to 10% ffmpeg4 to me, lower when flowblade and idjc are updated and mlt can be removed. Should just be old or dead packages left.

**probably 4 'forever'**

- vlc 4.x release (not ffmpeg4) is ultra slow
- olive should be axed i think
- qtox git is archived
- moc is old and probably dead
- kodi hasn't been touched in 10 years, not sure it compiles at all

**6 compatible**

- tvheadend release is slow
- qtav has another fork
- mediastreamer/linphone stuff i wanted to stay away from but probably possible
- deadbeef needs some newer packages for update, doable same as linphone
- hedgewars + ffmpeg6 built fine with https://github.com/hedgewars/hw/pull/74

What helps/hurts doing the big qt/gtk/electron stuff upfront vs after other packages that use ffmpeg6? Either way seems like no one will be able to update cleanly for a good while. More of a pain for the build process? Or do you see the ffmpeg PR being picked at and slowly added?

Should be able to get on irc this week if that's easier to talk over. Recovering from massive jetlag.

**Comment by zlice** [comment]
Made ffmpeg ffmpeg4 and ffmpeg6 a new package. Haven't built yet, don't know when I can test.

shiboken/pyside are semi unrelated now that qt and webkit stuff are dropped. Can make a new PR for that if it matters.

**Comment by zlice** [comment]
obs needed `qt6-base-private-devel` ? not sure what that is in 6.7.2

audacity update pending https://github.com/void-linux/void-packages/pull/50047 (either works, have patch for annoying scrollbar bug)

everything else built fine

idjc and flowblade updates can move those to 6

still need to test install and en/decoding

**Comment by chrysos349** [comment]
hey, great job here! do you plan to update ffmpeg to 7.0.1?

**Comment by classabbyamp** [comment]
it would make more sense to leave the ffmpeg 4 binaries as-is for now, and add the ffmpeg 6 binaries as <binname>6 to avoid conflicts. once most things support ffmpeg 6 and ffmpeg = 6.x, then we can change the default binary names

**Comment by zlice** [comment]
> it would make more sense to leave the ffmpeg 4 binaries as-is for now, and add the ffmpeg 6 binaries as 6 to avoid conflicts. once most things support ffmpeg 6 and ffmpeg = 6.x, then we can change the default binary names

i don't see why anyone would want the ffmpeg6 libraries but ffmpeg4 binaries. you won't be able to use new codecs on command line. can you overwrite the binaries for 4 with 6? that was part of the original PR having 6 as the main `ffmpeg` package.

only 4 packages can't move past ffmpeg4 : vlc, mediastreamer(linphone), olive, moc

vlc and mediastreamer may eventually update but olive is probably a lost cause and moc hasn't been touched since 2016 or something. i think it makes more sense to update everything else, maybe poke tvheadend where the 'q1' 4.4 release is or use the recommended 'dev' version - and leave vlc/mediastreamer on 4 as a library only. everything will just end up being build more otherwise won't it?

- updated digikam/qtav - qtav was absorbed and patched in digikam 8.0.0
- updated deadbeef finally

tested both on my old branch with 6 as main...looks like qt5-webengine broke over the last month or so on some chromium re2 StringPiece class

**Comment by classabbyamp** [comment]
my comment got messed up by markdown, it should say `<binname>6` not just 6

**Comment by ahesford** [comment]
Do any of the remaining v4 dependents actually require command-line tools? If not, I think making the original `ffmpeg` a library-only package is fine and we can let `ffmpeg6` be the only provider of command-line tools (and without versions in the executable names). But if VLC *et al.* require command-line tools, we probably need to take advice from @classabbyamp and let the new package provide tools with `6` at the end of their names.

**Comment by zlice** [comment]
> But if VLC _et al._ require command-line tools, we probably need to take advice from @ classabbyamp and let the new package provide tools with `6` at the end of their names.

afaik they don't. didn't look at olive code but the others look like code and no scripts calling anything

did get kodi 21 going, but xbps complains that it makes a `/kodi` instead of putting some python scripts for ps3 remote in `/opt` or somewhere else. not sure how to handle that short of moving them under `/usr/share`

**Comment by classabbyamp** [comment]
would it be possible to split this into smaller PRs, e.g.:
1. new package: ffmpeg6
2. opencv update
3. smaller batches of ffmpeg 4->6 migrations
4. big packages like chromium or webengine on their own (or held off until there's an update for them)

as it stands, it's very hard to review because of the size

**Comment by zlice** [comment]
it is, ffmpeg touches everything.

i can split more PRs, but some rebuilds will dupe ffmpeg revdeps

if you have any ideas for priority on waves of packages to make things smoother lmk

**updates**

- guvcview: update to 2.0. (dupe revbump for ffmpeg6)
- handbrake: update to 1.6. (dupe revbump)
- opencv: update to 4.7. (retroshare dupe revbump)
- unpaper: update to 7.0. (dupe revbump)
- deadbeef: update to 1.9. (libdispatch, dupe revbump)
- libwaylandpp: update to 1.0 / kodi: update to 21.0 (kodi dupe revbump)
- spek relacement (dupe revbump)
- audacity: update to 3.5. (another PR has this)
- idjc (another PR)
- flowblade (another PR)

**big packages**

- gtk4
- qt5
- qt5-webengine
- qt6-pdf (webengine)
- qt6-multimedia
- konqeror
- chromium
- electron19
- electron24

**misc**

- qtav/digikam (should be no extra work as a separate PR)

**Comment by zlice** [comment]
https://github.com/void-linux/void-packages/pull/51123

https://github.com/void-linux/void-packages/pull/51124

only half of opencv passed ci for some reason, seems fine locally for me

**Comment by zlice** [comment]
https://github.com/void-linux/void-packages/pull/51132

still probably going to use this branch/pr for tracking

**Comment by Pestdoktor** [comment]
Just wanted to report in that I haven't had an issue running ffmpeg and dependencies off this branch (as pulled at the end of march). Great work!

**Comment by zlice** [comment]
will do, probably later tonight. ill probably sort by stuff i use reguarly that has been working fine with 6, version update packages separately(?), then by w/e groups that make sense. hold on the big electron/gtk/qt stuff like you said and wait for those to update and switch ffmpeg.

**Comment by classabbyamp** [comment]
sounds good to me, no rush too

**Comment by classabbyamp** [comment]
by my reckoning, this is all that's left in this PR:

- _New package: vpl-gpu-rt-24.2.5_ - #51133 
- _python3-pyside2: revbump for qt5-5.15.11_ and _shiboken2: revbump for qt5-5.15.11_ - should be their own PR, not sure why they're in here
- _audacity: update to 3.5.1_ - #50047 or #51407

Thus, this can be closed now.

other packages I see in the repos still on ffmpeg4:
- electron19
- electron24
- qt5
- qt5-webengine
- qt6-pdf
- qt6-multimedia
- vlc
- qtox
- moc
- mpv-mpris
- tvheadend
- opencv
- olive
- mediastreamer

**Comment by zlice** [comment]
sounds right

pyside/shiboken was due to the original having ffmpeg6 as main and rebuilding qt i think. will take another look at those and make a pr if theyre still not compiling currently.

tvheadend still no 4.4 release, the 4.3 'dev' works. electron and qt should be the only things left then. others are on ffmpeg4 until question-marks

thanks for all the assistance abby

**Comment by classabbyamp** [comment]
made a tracking issue: #51522 

can you make a PR for the qt5-related patches?

**Pull request closed without merging**
